server/svix-server/src/v1/endpoints/attempt.rs

```diff
@@ -65,6 +65,112 @@ impl From<messageattempt::Model> for MessageAttemptOut {
+    created_at: DateTimeWithTimeZone,
+    status: MessageStatus,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    next_attempt: Option<DateTimeWithTimeZone>,
+}
+
+impl AttemptedMessageOut {
+    pub fn from_dest_and_msg(
+        dest: messagedestination::Model,
+        msg: message::Model,
+    ) -> AttemptedMessageOut {
+        AttemptedMessageOut {
+            event_type: msg.event_type,
+            event_id: msg.uid,
+            channels: msg.channels,
+            payload: msg.payload,
+            id: msg.id,
+            created_at: msg.created_at,
+            status: dest.status,
+            next_attempt: dest.next_attempt,
+        }
+    }
+}
+
+/// Additional parameters (besides pagination) in the query string for the "List Attempted Messages"
+/// enpoint.
+#[derive(Debug, Deserialize, Validate)]
+pub struct ListAttemptedMessagesQueryParameters {
+    #[validate]
+    channel: Option<EventChannel>,
+    status: Option<MessageStatus>,
+}
+
+/// Fetches a list of [`AttemptedMessageOut`]s associated with a given app and endpoint.
+async fn list_attempted_messages(
+    Extension(ref db): Extension<DatabaseConnection>,
+    ValidatedQuery(mut pagination): ValidatedQuery<Pagination<MessageId>>,
+    ValidatedQuery(ListAttemptedMessagesQueryParameters { channel, status }): ValidatedQuery<
+        ListAttemptedMessagesQueryParameters,
+    >,
+    Path((_app_id, endp_id)): Path<(ApplicationId, EndpointId)>,
+    AuthenticatedApplication {
+        permissions: _,
+        app: _,
+    }: AuthenticatedApplication,
+) -> Result<Json<ListResponse<AttemptedMessageOut>>> {
+    let limit = pagination.limit;
+    let iterator = pagination.iterator.take();
+
+    let mut dests_and_msgs = messagedestination::Entity::secure_find_by_endpoint(endp_id)
+        .find_also_related(message::Entity)
+        .order_by_desc(messagedestination::Column::CreatedAt)
+        .limit(limit + 1);
+
+    if let Some(iterator) = iterator {
+        dests_and_msgs = dests_and_msgs.filter(messagedestination::Column::MsgId.lt(iterator));
+    }
+
+    if let Some(channel) = channel {
+        dests_and_msgs = dests_and_msgs.filter(message::Column::Channels.contains(&channel));
+    }
+
+    if let Some(status) = status {
+        dests_and_msgs = dests_and_msgs.filter(messagedestination::Column::Status.eq(status));
+    }
+
+    Ok(Json(AttemptedMessageOut::list_response(
+        dests_and_msgs
+            .all(db)
+            .await?
+            .into_iter()
+            .map(
+                |(dest, msg): (messagedestination::Model, Option<message::Model>)| {
+                    let msg = msg.ok_or_else(|| {
+                        tracing::error!(
+                            "messagedestination::Enitity has no associated message::Entity"
+                        );
+                        Error::Database("Malformed data".to_owned())
+                    })?;
+
+                    Ok(AttemptedMessageOut::from_dest_and_msg(dest, msg))
```

Comment:
Does this actually work? I don't understand how the error is propagated outside of the closure.

Also, I think doing:
```suggestion
                    AttemptedMessageOut::from_dest_and_msg(dest, msg.expect("Message missing for message endpoint exist"))
```
Is sufficient, as Axum automatically catches unwraps and returns a 500 (I think, check me!).

Though it's good either way, just wanted to keep our code concise.